### PR TITLE
feature: Add TypeChat intent detection

### PR DESCRIPTION
**Pull request by JasonYeMSFT**
https://github.com/microsoft/vscode-azure-agent/issues/77

### CODE REVIEW

File: src/chat/followUpGenerator.ts [hunk 1]
```diff
@@ -19,7 +20,10 @@ export async function generateExtensionCommandFollowUpsOther(copilotContent: str
     const intentDetectionTargets = availableCommands
         .map((command) => ({ name: command.name, intentDetectionDescription: command.intentDescription || command.displayName }));
 
-    const detectedIntentionTarget = await detectIntent(intentDetectionTargets, copilotContentAgentRequest);
+    const detectedIntentionTarget = await detectIntentNaturalLanguage(intentDetectionTargets, copilotContentAgentRequest);
+    const detectedIntentionTargetViaTypeChat = await detectIntentTypeChat(intentDetectionTargets, copilotContentAgentRequest);
+    ext.outputChannel.appendLog(`detectedTargetViaNaturalLanguage: ${JSON.stringify(detectedIntentionTarget)}, detectedTargetViaTypeChat: ${JSON.stringify(detectedIntentionTargetViaTypeChat)}`);
```
Comment (JasonYeMSFT) [inline src/chat/followUpGenerator.ts:25]: The changes are ready for review. I'll find some time to run all the benchmarks and make sure the correctness doesn't deviate too far from the existing natural language approach.

File: src/chat/rag.ts [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 import { createHttpHeaders } from "@azure/core-rest-pipeline";
 import { ClientSecretCredential } from "@azure/identity";
+// @ts-ignore
```
Comment (MRayermannMSFT) [inline src/chat/rag.ts:8]: Why was this added?

File: src/chat/intentDetection.ts [hunk 3]
```diff
@@ -3,16 +3,54 @@
...
 export type IntentDetectionTarget = {
     name: string,
     intentDetectionDescription: string,
 }
 
-export async function detectIntent(targets: IntentDetectionTarget[], request: AgentRequest): Promise<IntentDetectionTarget | undefined> {
+/**
+ * Detect user intent using Type Chat.
+ */
+export async function detectIntentTypeChat(targets: IntentDetectionTarget[], request: AgentRequest): Promise<IntentDetectionTarget | undefined> {
```
Comment (MRayermannMSFT) [inline src/chat/intentDetection.ts:23]: To avoid changing files, let's keep an `export async function detectIntent` and have it call `detectIntentTypeChat`/`detectIntentNaturalLanguage`, and it can handle logging TypeChat vs old-way.

File: src/chat/intentDetection.ts [hunk 4]
```diff
@@ -3,16 +3,54 @@
...
+    const schema = getZodSchema(targets);
+    const validator = createZodJsonValidator(schema, "Action");
+    const translator = createJsonTranslator(typechatLanguageModel, validator);
+    const response = await translator.translate(request.userPrompt);
+    if (response.success) {
+        const data = response.data;
+        const intent = (data as any).intent;
+        // Note: Eventually the Zod schema will be translated to Typescript. We cannot use Typescript keywords such as "unknown", "undefinef", etc.
+        if (intent === "UnknownIntent") {
+            ext.outputChannel.appendLog(`Failed to detect user intent using typechat. The intent is unknown give the available targets.`);
```
Comment (MRayermannMSFT) [inline src/chat/intentDetection.ts:33]: Are you going to leave the logging in when you merge? If so, can you just use `console.log`? I've been using that for other debugging purposes. Maybe you can show me this `ext.outputChannel` too though, maybe it's a better choice?

File: src/chat/slashCommands.ts [hunk 5]
```diff
@@ -186,7 +187,10 @@ export class SlashCommandsOwner implements IAgentRequestHandler {
             if (!result && prompt !== "" && this._disableIntentDetection !== true) {
                 const intentDetectionTargets = Array.from(this._invokeableSlashCommands.entries())
                     .map(([name, config]) => ({ name: name, intentDetectionDescription: config.intentDescription || config.shortDescription }));
-                const detectedTarget = await detectIntent(intentDetectionTargets, request);
+                const detectedTarget = await detectIntentNaturalLanguage(intentDetectionTargets, request);
+                const detectedTargetViaTypeChat = await detectIntentTypeChat(intentDetectionTargets, request);
+                ext.outputChannel.appendLog(`detectedTargetViaNaturalLanguage: ${JSON.stringify(detectedTarget)}, detectedTargetViaTypeChat: ${JSON.stringify(detectedTargetViaTypeChat)}`);
```
Comment (MRayermannMSFT) [inline src/chat/slashCommands.ts:192]: Ya, move this to be internal like I said in other comment.

File: src/chat/typechat/result.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+export function error(message: string): Error {
+    return { success: false, message };
+}
+
+/**
+ * Obtains the value associated with a successful `Result<T>` or throws an exception if
+ * the result is an error.
+ * @param result The `Result<T>` from which to obtain the `data` property.
+ * @returns The value of the `data` property.
+ */
```
Comment (MRayermannMSFT) [inline src/chat/typechat/result.ts:39]: Uh missing code below this?

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { PromptSection, Result, error, success } from "typechat";
+import * as vscode from "vscode";
+
+export type CopilotInteractionResult = { copilotResponded: true, copilotResponse: string } | { copilotResponded: false, copilotResponse: undefined };
+
+export const typechatLanguageModel = {
+    async complete(prompt: PromptSection[]): Promise<Result<string>> {
+        // @todo: Experiement with more advanced usage of prompts
+        // such as: including history, using both system and user prompts, etc.
+        const promptText = prompt[0].content;
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:10]: So is the `promp` parameter coming from TypeChat? If so, let's use it as intened and pass all messages into VS Code's LLM.

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { PromptSection, Result, error, success } from "typechat";
+import * as vscode from "vscode";
+
+export type CopilotInteractionResult = { copilotResponded: true, copilotResponse: string } | { copilotResponded: false, copilotResponse: undefined };
+
+export const typechatLanguageModel = {
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:6]: Please be consistent with casign of TypeChat (in this case I guess I'd expect `typeChatLanguageModel`?

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    async complete(prompt: PromptSection[]): Promise<Result<string>> {
+        // @todo: Experiement with more advanced usage of prompts
+        // such as: including history, using both system and user prompts, etc.
+        const promptText = prompt[0].content;
+        const messages = [
+            new vscode.LanguageModelChatUserMessage(promptText)
+        ];
+        const cancellationTokenSource = new vscode.CancellationTokenSource();
+        try {
+            const response = await vscode.lm.sendChatRequest(
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:16]: Please reuse code from `copilotInteractions.ts`. I'm think you'll want to do something like:
1. Have something like `getTypeChatLanguageModel` which takes the `AgentRequest` which is associated with the current invocation of the agent
2. In the `complete` function which the getter creates, when you call something from `copilotInteractions`, you will use the properties from the `AgentRequest` passed into `getTypeChatLanguageModel` which creates a new one with the prompt/history TypeChat wants used.

File: src/chat/typechat/zodSchemaHelper.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { z } from "zod";
```
Comment (MRayermannMSFT) [inline src/chat/typechat/zodSchemaHelper.ts:1]: This file seems less like "zod schema helper" and more like "zod intent detection helper". Should this code should just live in `intentDetection.ts`?

File: src/chat/typechat/result.ts
Comment (bwateratmsft) [inline src/chat/typechat/result.ts:1]: With the TypeChat package being included as a tarball in the source, is this file still needed?

File: src/chat/typechat/zodSchemaHelper.ts (hunk 10 above)
Comment (JasonYeMSFT) [reply src/chat/typechat/zodSchemaHelper.ts:1]: I prefer keeping Zod related logic outside `intentDetection.ts`. I think `zodIntentDetectionHelper` is a better name for this file.

File: src/chat/typechat/vscodeTypechatModel.ts (hunk 7 above)
Comment (JasonYeMSFT) [reply src/chat/typechat/vscodeTypechatModel.ts:10]: Sure.

File: src/chat/rag.ts (hunk 2 above)
Comment (JasonYeMSFT) [reply src/chat/rag.ts:8]: Because "@azure/openai"'s type issue will generate an error during compilation. I can get rid of the ts-ignore though since it doesn't seem to block webpack from emitting bundled code.

File: src/chat/typechat/result.ts
Comment (JasonYeMSFT) [reply src/chat/typechat/result.ts:1]: You are right. This file is not needed now.

File: src/chat/typechat/result.ts (hunk 6 above)
Comment (JasonYeMSFT) [reply src/chat/typechat/result.ts]: The file is gone now.

File: src/chat/typechat/vscodeTypechatModel.ts (hunk 8 above)
Comment (JasonYeMSFT) [reply src/chat/typechat/vscodeTypechatModel.ts]: I renamed all Typechat to TypeChat.

File: src/chat/intentDetection.ts (hunk 4 above)
Comment (JasonYeMSFT) [reply src/chat/intentDetection.ts:35]: I moved the logging to the private place using console.log.

File: src/chat/typechat/vscodeTypechatModel.ts (hunk 9 above)
Comment (JasonYeMSFT) [reply src/chat/typechat/vscodeTypechatModel.ts:93]: I think I have implemented what you had asked for.

File: src/chat/typechat/zodIntentDetectionHelper.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    intentDetectionDescription: z.ZodLiteral<string>;
+}
+
+export type ZodIntentDetectionResult = {
+    intent: z.ZodEnum<[string, ...string[]]>;
+};
+
+export type ZodIntentDetectionSchema = ZodIntentDetectionTarget | ZodIntentDetectionResult;
+
+export function getZodSchema(targets: IntentDetectionTarget[]): { [key: string]: z.ZodObject<ZodIntentDetectionSchema> } {
```
Comment (MRayermannMSFT) [inline src/chat/typechat/zodIntentDetectionHelper.ts:15]: Nit: rename to like `getZodIntentDetectionSchema`?

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
+import { error, success, type PromptSection, type Result, type TypeChatLanguageModel } from "typechat";
+import * as vscode from "vscode";
+import { type AgentRequest } from "../agent";
+import { getResponseAsStringCopilotInteraction } from "../copilotInteractions";
+
+export type CopilotInteractionResult = { copilotResponded: true, copilotResponse: string } | { copilotResponded: false, copilotResponse: undefined };
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:6]: This shouldn't be needed. This type exists elsewhere.

File: src/chat/intentDetection.ts [hunk 13]
```diff
@@ -3,16 +3,57 @@
...
+            return undefined;
+        } else {
+            for (const target of targets) {
+                if (target.name === intent) {
+                    return target;
+                }
+            }
+            ext.outputChannel.appendLog(`Failed to detect user intent using typechat. The intent ${intent} was not one of the expected targets.`);
```
Comment (MRayermannMSFT) [inline src/chat/intentDetection.ts:44]: Two follow up PRs I'd like to see:
1. Refactor `detectIntentTypeChat` and `detectIntentNaturalLanguage` to return a `failureReason` and just log that in `detectIntent`
2. Extension wide, swap our usages of `console.log` to `ext.outputChannel.appendLog` (or ideally the newer version that has verbosity levels)

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+export function getTypeChatLanguageModel(request: AgentRequest): TypeChatLanguageModel {
+    // TypeChat doesn't use the system prompt so leave it blank.
+    const systemPrompt = "";
+    const command = request.command;
+    // @todo: Pull the name and the id of the extension
+    const participant = { extensionId: "", name: "" };
+    return {
+        // Before calling TypeChat, we have already summarized the conversation history including the latest user prompt.
+        // TypeChat will call this function with a further augmented user prompt and optionally assistant prompts if it needs to repair the response.
+        // If there are multiple prompts, the first prompt will be the earliest prompt and the prompts will be interleaving user prompts and assistance prompts.
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:20]: Can you say oldest instead of earliest?

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        // e.g. [1st_user_prompt, 1st_assistance_prompt (aka. Response of the 1st_user_prompt), 2nd_user_prompt]
+        async complete(prompt: string | PromptSection[]): Promise<Result<string>> {
+            let newRequest: AgentRequest;
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:23]: Nit, name it `typeChatRequest`?

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            console.log("Prompt generated by TypeChat", prompt);
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:24]: This shouldn't be needed. Because you are using copilotInteractions, then it will be logged if logging is turned on. :)

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+export type CopilotInteractionResult = { copilotResponded: true, copilotResponse: string } | { copilotResponded: false, copilotResponse: undefined };
+
+/**
+ * @param request Original user prompt and its context.
+ */
+export function getTypeChatLanguageModel(request: AgentRequest): TypeChatLanguageModel {
+    // TypeChat doesn't use the system prompt so leave it blank.
+    const systemPrompt = "";
+    const command = request.command;
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:14]: I would just leave command empty.

File: src/chat/typechat/vscodeTypechatModel.ts [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+                newRequest = {
+                    ...request,
+                    // userPrompt.content should never be empty but handle it just in case
+                    userPrompt: userPrompt?.content ?? "",
+                    context: {
+                        history: historyTurns
+                    }
+                };
+            }
+            console.log("NewRequest forged for TypeChat", newRequest);
```
Comment (MRayermannMSFT) [inline src/chat/typechat/vscodeTypechatModel.ts:62]: Just get rid of this.

File: src/chat/intentDetection.ts (hunk 13 above)
Comment (JasonYeMSFT) [reply src/chat/intentDetection.ts]: I removed the log output for now and added two todo comments.

**Review by MRayermannMSFT — APPROVED** [review]

File: src/chat/intentDetection.ts [hunk 19]
```diff
@@ -3,16 +3,55 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import { createJsonTranslator } from "typechat";
+import { createZodJsonValidator } from "typechat/zod";
 import { type AgentRequest } from "./agent";
 import { getResponseAsStringCopilotInteraction, getStringFieldFromCopilotResponseMaybeWithStrJson } from "./copilotInteractions";
 import { summarizeHistoryThusFar } from "./summarizing";
+import { getTypeChatLanguageModel } from "./typechat/vscodeTypeChatModel";
```
[the remark this answers is not in the document]
Comment (JasonYeMSFT) [reply src/chat/intentDetection.ts]: Done.
